app/services/themes/create.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+class Themes::Create
+  include Service::Base
+
+  params do
+    attribute :name, :string
+    attribute :user_id, :integer
+    attribute :user_selectable, :boolean, default: false
+    attribute :color_scheme_id, :integer
+    attribute :component, :boolean, default: false
+    attribute :theme_fields, :array
+    attribute :default, :boolean, default: false
+  end
+
+  policy :ban_in_allowlist_mode
+
+  step :initialize_theme
+
+  policy :ban_for_remote_theme
+
+  step :set_theme_fields
+
+  transaction do
+    step :save_theme
+    step :update_default_theme
+    step :log_theme_change
+  end
+
+  private
+
+  def ban_in_allowlist_mode
+    Theme.allowed_remote_theme_ids.nil?
+  end
+
+  def initialize_theme(params:)
+    context[:theme] = Theme.new(
+      name: params.name,
+      user_id: params.user_id,
+      user_selectable: params.user_selectable,
+      color_scheme_id: params.color_scheme_id,
+      component: params.component,
+    )
+  end
+
+  def ban_for_remote_theme(params:, theme:)
+    return true if params.theme_fields.blank?
+    !theme.remote_theme&.is_git?
+  end
+
+  def set_theme_fields(params:, theme:)
+    return if params.theme_fields.blank?
+
+    params.theme_fields.each do |field|
+      theme.set_field(
+        target: field[:target],
+        name: field[:name],
+        value: field[:value],
+        type_id: field[:type_id],
+        upload_id: field[:upload_id],
+      )
+    end
+  end
+
+  # TODO (martin) Ask loic about this, the old theme controller expected the
+  # errors from the theme model save to be shown to the user.
+  def save_theme(theme:)
+    theme.save!
```

Comment:
For now, either use a `rescue` block and call `fail!` inside it or just call `save` and check for any errors (and call `fail!` :grin:), as we discussed IIRC.

It seems we should be able to do something like that in the `model` step, though:
```ruby
Theme.create(params.slice(:name, :user_id, :user_selectable, :color_scheme_id, :component)) do |theme|
  params.theme_fields.each do |field|
    theme.set_field(**field)
  end
end
```
The only problem is still that `Theme#set_field` raises because it does validations that should be done in `ThemeField`. Maybe a small refactor (moving validations where they should happen) could resolve that situation.

Maybe that’s an approach we could apply for updates too (`Model.find_by(…).tap {|model| model.update(…) }`), I don’t know :thinking: 

Otherwise, we could implement a new step, something like `persist_model :my_model`, it would call `save` on it, check if the model is both valid and persisted, if that’s not the case, it would stop the execution flow.

Reply:
Sure I could try a refactor of `set_field`, and otherwise do the rescue/error check for save. persist_model sounds nice too :+1:

Reply:
I just went for the theme.save/fail! model errors for now

Reply:
I’m working on adding a new step to handle exceptions: https://github.com/discourse/discourse/pull/29660.
That would probably come in handy in a situation like this :grin: